java/ql/src/Security/CWE/CWE-338/JHipsterGeneratedPRNG.ql

```diff
@@ -0,0 +1,48 @@
+/**
+ * @name Detect JHipster Generator Vulnnerability CVE-2019-16303
+ * @description Detector for the CVE-2019-16303 vulnerability that existed in the JHipster code generator.
+ * @kind problem
+ * @problem.severity error
+ * @precision very-high
+ * @id java/jhipster-prng
+ * @tags security
+ *       external/cwe/cwe-338
+ */
+
+import java
+import semmle.code.java.frameworks.apache.Lang
+
+private class PredictableApacheRandomStringUtilsMethod extends Method {
+  PredictableApacheRandomStringUtilsMethod() {
+    this.getDeclaringType() instanceof TypeApacheRandomStringUtils
+  }
+}
+
+private class PredictableApacheRandomStringUtilsMethodAccess extends MethodAccess {
+  PredictableApacheRandomStringUtilsMethodAccess() {
+    this.getMethod() instanceof PredictableApacheRandomStringUtilsMethod and
+    // The one valid use of this type that uses SecureRandom as a source of data.
+    not this.getMethod().getName() = "random"
```

Comment:
I don't really understand why you exclude all methods named `random`. As far as I can tell they are all bad except the one that takes a `Random` as parameter which allows for passing it a `SecureRandom` instance.

Reply:
You could exclude this method by writing something like: 
```
not (this.getMethod().hasName("random") and this.getMethod().getParameter(6).getType() instanceof TypeRandom)
```
If you want to be even more fancy you could check whether the type (bound) of argument `6` matches a secure random type (probably overkill).

Reply:
As far as I'm aware, JHipster never generated any code that used the `random` method.